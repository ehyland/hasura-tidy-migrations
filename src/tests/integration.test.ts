import {
  setupWorkspace,
  cleanupWorkspace,
  runCli,
  RELATIVE_WORKSPACE_MIGRATIONS,
  stdoutToUpdatedMessages,
  FunctionResolveType,
  WORKSPACE_MIGRATIONS,
} from './helpers';

import path from 'path';
import fs from 'fs-extra';

const readWorkspaceMigration = (name: string) =>
  Promise.all([
    fs.readFile(path.resolve(WORKSPACE_MIGRATIONS, name, 'up.yaml'), 'utf8'),
    fs.readFile(path.resolve(WORKSPACE_MIGRATIONS, name, 'down.yaml'), 'utf8'),
  ]);

describe('integration', () => {
  beforeAll(async () => {
    await setupWorkspace();
  });

  afterAll(async () => {
    await cleanupWorkspace();
  });

  describe('when run on dirty migrations', () => {
    let out: FunctionResolveType<typeof runCli>;

    beforeAll(async () => {
      out = await runCli(RELATIVE_WORKSPACE_MIGRATIONS);
    });

    it('when called with <migrations-dir>', async () => {
      expect(out.stderr).toMatchInlineSnapshot(`""`);
      expect(stdoutToUpdatedMessages(out.stdout)).toEqual([
        'Updated 1579996071502_create_table_public_users/up.yaml',
        'Updated 1580088801913_create_table_public_todos/up.yaml',
        'Updated 1580089036643_set_fk_public_todos_user_id/down.yaml',
        'Updated 1580089036643_set_fk_public_todos_user_id/up.yaml',
        'Updated 1580089400009_run_sql_migration/up.yaml',
      ]);
    });

    it('formats sql in migrations', async () => {
      const setFkPublicTodosUserId = await readWorkspaceMigration(
        '1580089036643_set_fk_public_todos_user_id'
      );
      const createTablePublicTodos = await readWorkspaceMigration(
        '1580088801913_create_table_public_todos'
      );
      const runSqlMigration = await readWorkspaceMigration('1580089400009_run_sql_migration');
      const createTablePublicUsers = await readWorkspaceMigration(
        '1579996071502_create_table_public_users'
      );
      const addRelationshipTablePublicTodos = await readWorkspaceMigration(
        '1580089123700_add_relationship__table_public_todos'
      );

      [
        ...setFkPublicTodosUserId,
        ...createTablePublicTodos,
        ...runSqlMigration,
        ...createTablePublicUsers,
        ...addRelationshipTablePublicTodos,
      ].forEach(migration => {
        expect(migration).toMatchSnapshot();
      });
    });

    describe('when run on clean migrations', () => {
      beforeAll(async () => {
        out = await runCli(RELATIVE_WORKSPACE_MIGRATIONS);
      });

      it('does not update files', () => {
        expect(out.stderr).toMatchInlineSnapshot(`""`);
        expect(stdoutToUpdatedMessages(out.stdout)).toEqual([]);
      });
    });
  });
});
