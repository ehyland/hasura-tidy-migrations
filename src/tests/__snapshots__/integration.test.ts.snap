// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration when run on dirty migrations formats sql in migrations 1`] = `
"- args:
    sql: |-
      alter table
        \\"public\\".\\"todos\\"
      add
        constraint \\"todos_user_id_fkey\\" foreign key (\\"user_id\\") references \\"public\\".\\"users\\" (\\"id\\") on update restrict on delete restrict;
  type: run_sql
"
`;

exports[`integration when run on dirty migrations formats sql in migrations 2`] = `
"- args:
    sql: |-
      alter table
        \\"public\\".\\"todos\\" drop constraint \\"todos_user_id_fkey\\"
  type: run_sql
"
`;

exports[`integration when run on dirty migrations formats sql in migrations 3`] = `
"- args:
    sql: |-
      CREATE TABLE \\"public\\".\\"todos\\"(
        \\"id\\" serial NOT NULL,
        \\"title\\" text NOT NULL,
        \\"is_completed\\" boolean NOT NULL DEFAULT false,
        \\"is_public\\" boolean NOT NULL DEFAULT false,
        \\"created_at\\" timestamptz NOT NULL DEFAULT now(),
        \\"user_id\\" text NOT NULL,
        PRIMARY KEY (\\"id\\")
      );
  type: run_sql
- args:
    name: todos
    schema: public
  type: add_existing_table_or_view
"
`;

exports[`integration when run on dirty migrations formats sql in migrations 4`] = `
"- args:
    sql: DROP TABLE \\"public\\".\\"todos\\"
  type: run_sql
"
`;

exports[`integration when run on dirty migrations formats sql in migrations 5`] = `
"- args:
    cascade: false
    sql: |-
      CREATE
      OR REPLACE VIEW \\"public\\".\\"online_users\\" AS
      SELECT
        users.id,
        users.last_seen
      FROM
        users
      WHERE
        (
          users.last_seen >= (now() - '00:00:30' :: interval)
        );
  type: run_sql
- args:
    name: online_users
    schema: public
  type: add_existing_table_or_view
"
`;

exports[`integration when run on dirty migrations formats sql in migrations 6`] = `
"[]
"
`;

exports[`integration when run on dirty migrations formats sql in migrations 7`] = `
"- args:
    sql: |-
      CREATE TABLE \\"public\\".\\"users\\"(
        \\"id\\" text NOT NULL,
        \\"name\\" text NOT NULL,
        \\"created_at\\" timestamptz NOT NULL DEFAULT now(),
        \\"last_seen\\" timestamptz,
        PRIMARY KEY (\\"id\\")
      );
  type: run_sql
- args:
    name: users
    schema: public
  type: add_existing_table_or_view
"
`;

exports[`integration when run on dirty migrations formats sql in migrations 8`] = `
"- args:
    sql: DROP TABLE \\"public\\".\\"users\\"
  type: run_sql
"
`;

exports[`integration when run on dirty migrations formats sql in migrations 9`] = `
"- args:
    name: user
    table:
      name: todos
      schema: public
    using:
      foreign_key_constraint_on: user_id
  type: create_object_relationship
"
`;

exports[`integration when run on dirty migrations formats sql in migrations 10`] = `
"- args:
    relationship: user
    table:
      name: todos
      schema: public
  type: drop_relationship
"
`;
